### PR TITLE
Outgoing ddi rules

### PR DESCRIPTION
**Pull request by Kaian**
Add two new tables for OutgoingDDIRules and OutgoingDDIRulePatterns.

Users can now assign a outgoingDDIRule that will determine
the final DDI presented based on the destination number
and the matching patterns in OutgoingDDIRulePatterns.

This PR implements #192

### CODE REVIEW

**Comment by Kaian** [comment]
Should company have default outgoing ddi rules?

They have default outgoing ddi...

**Comment by cruzccl** [comment]
It would be nice as it would enable something like "all the users calling to XXXXX (or to Y country) will show ZZZZZZ instead of their normal number" without having to add this rule for all the users (and new users, that will be forgotten for sure).

Collision between user rules and company rules should be solved as usual: if a user rule applies, use it. If not, evaluate company rules.

Nice feature!

File: library/IvozProvider/Mapper/Sql/OutgoingDDIRules.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+class OutgoingDDIRules extends Raw\OutgoingDDIRules
+{
+    protected function _save(\IvozProvider\Model\Raw\OutgoingDDIRules $model,
+            $recursive = false, $useTransaction = true, $transactionTag = null, $forceInsert = false
+    )
+    {
+        $nullableFields = array(
+                "force"        => "forcedDDIId",
+        );
+        $routeType = $model->getDefaultAction();
```
Comment (Kaian) [inline library/IvozProvider/Mapper/Sql/OutgoingDDIRules.php]: `s/routeType/defaultAction/`

File: library/IvozProvider/Mapper/Sql/OutgoingDDIRulesPatterns.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+class OutgoingDDIRulesPatterns extends Raw\OutgoingDDIRulesPatterns
+{
+    protected function _save(\IvozProvider\Model\Raw\OutgoingDDIRulesPatterns $model,
+            $recursive = false, $useTransaction = true, $transactionTag = null, $forceInsert = false
+    )
+    {
+        $nullableFields = array(
+                "force"        => "forcedDDIId",
+        );
+        $routeType = $model->getAction();
```
Comment (Kaian) [inline library/IvozProvider/Mapper/Sql/OutgoingDDIRulesPatterns.php]: `s/routeType/action/`

**Comment by Kaian** [comment]
I think this should be merged after #197 and replace the regexp field in OutgoingDDIRulePatterns with matchListId so all numeric regular expressions are in the same entity.

This is a bit less _confortable_ for company administrators as they have to also create matchlist for OutgoingRulesPatterns but will about duplicity of data and process (where the same logic of _number matches regexp_ is spread in multiples files)

**Comment by Kaian** [comment]
Blocked by #201

File: library/IvozProvider/Model/Raw/Companies.php [hunk 3]
```diff
@@ -3835,6 +3847,96 @@ public function getMusicOnHold($where = null, $orderBy = null, $avoidLoading = f
...
     /**
+     * Sets dependent relations OutgoingDDIRules_ibfk_1
+     *
+     * @param array $data An array of \IvozProvider\Model\Raw\OutgoingDDIRules
+     * @return \IvozProvider\Model\Raw\Companies
+     */
+    public function setOutgoingDDIRules(array $data, $deleteOrphans = false)
+    {
+        if ($deleteOrphans === true) {
+
```
Comment (mmadariaga) [inline library/IvozProvider/Model/Raw/Companies.php]: I think that we'll need to set some coding standards for white spaces :)

File: portals/application/modules/rest/controllers/OutgoingDDIRulesPatternsController.php [hunk 4]
```diff
@@ -34,14 +34,22 @@ public function init()
      * @ApiReturn(type="object", sample="[{
      *     'id': '', 
      *     'outgoingDDIRuleId': '', 
+<<<<<<< HEAD
```
Comment (mmadariaga) [inline portals/application/modules/rest/controllers/OutgoingDDIRulesPatternsController.php]: Review this please

File: library/IvozProvider/Model/Raw/Companies.php (hunk 3 above)
Comment (Kaian) [inline library/IvozProvider/Model/Raw/Companies.php]: But that code is autogenerated by run-generators phing task

**Comment by Kaian** [comment]
I have update all source to use MatchLists instead of Regular expressions.

Also added OutgoingDDIRules at company level to make a global outgoing ddi selection (that can be overrider by user configuration).

@cruzccl 
You're example should be possible with actual implementation of outgoing ddi rules
- Create N matchlists: french, UK, ..., partners' cellphones
- Create an outgoing rule set and add one rule for each matchlist with the proper action (force french DDI, UK, and so on).

I think multiple rules will make thing more complicated, and current status covers the most common cases.

**Comment by Kaian** [comment]
I must also add that this feature **does NOT work with diverted calls** until #89 is implemented.

The main reason is that all _number_ input fields of the portals doesn't have an associated country, so its value has to pass for all the _preferred_ formating logics (which I prefer to avoid).

**Comment by Kaian** [comment]
Added documentation

File: library/IvozProvider/Model/Raw/Companies.php (hunk 3 above)
Comment (mmadariaga) [inline library/IvozProvider/Model/Raw/Companies.php]: Thats right :sweat:

**Pull request merged**